Add more onvif PTZ move modes

olijouve opened the pull request:
## Description:
Adding support for other PTZ move modes of http://www.onvif.org/ver10/ptz/wsdl specs

Onvif integration only supports RelativeMove where it should also supports AbsoluteMove and ContinuousMove.

For exemple chinese Goke GK7102 based IP camera only support ContinuousMove mode.

This PR add those new modes with avaibility to select mode and params in service call.

**Breaking change:**
The onvif camera service `camera.onvif_ptz` has been moved from the camera domain to the onvif domain.  onvif_ptz service was also renamed to ptz.
So service calls need to be updated to `onvif.ptz`.

**Changes:**
Add ContinuousMove mode
Add continuous_duration delay to stop movement after that delay
Add AbsoluteMove mode
Use service.data[key] instead of service.data.get(key) for keys with default schema values

**Lovelace integration sample:**
```  
  - type: entity-button 
      entity: camera.foscam
      icon: mdi:arrow-left-bold-outline
      show_name: false
      tap_action:
        action: call-service
        service: onvif.ptz
        service_data:
          entity_id: camera.foscam
          pan: LEFT
          tilt:
          zoom:
          move_mode: ContinuousMove
          continuous_duration: 0.8
          distance: 1
```

**Related issue (if applicable):** 
At least #27744 

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
https://github.com/home-assistant/home-assistant.io/pull/11557
Breaking change: Domain moved from camera to onvif so service calls need to be update to onvif.onvif_ptz 
If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

olijouve committed:
Adding support for PTZ move modes

Adding support for other PTZ move modes.
Onvif intergration used to only support RelativeMove where it should also supports AbsoluteMove, ContinuousMove and Stop.
For exemple Goke GK7102 based IP camera only support ContinuousMove mode.

This commit add those new modes with avaibility to select mode and params in service call.

olijouve committed:
Adding support for PTZ move modes

Adding support for other PTZ move modes.
Onvif intergration used to only support RelativeMove where it should also supports AbsoluteMove, ContinuousMove and Stop.
For exemple Goke GK7102 based IP camera only support ContinuousMove mode.

Update service helper for new avaibility to select mode and params in service call.

olijouve committed:
RelativeMode as default move_mode to avoid breakchange

RelativeMode as default move_mode to avoid breakchange

olijouve committed:
add missing attribute

add missing continuous_duration attribute

olijouve committed:
change service attribute label for continuous_duration

olijouve committed:
update description

fix wrong assertion for move_mode attr description

CiquattroFPV commented:
as stated in the issue opened by me https://github.com/home-assistant/home-assistant/issues/27744 I have tried this commit and the PTZ command are working for me.

olijouve committed:
Update services.yaml

REALogics commented:
@olijouve  

Using HS:DeveloperTools:Services and Foscam FI9928P ONVIF camera:

ZOOM DOES NOT work.  

PAN and TILT direction functions DO work with **distance** and **continuous_duration** parameters.  The **speed** parameter has no affect on speed.  The **continuous_duration** parameter only works up to max of 1.0 second, yet parameter description suggests plural **delay in seconds**.  Entering **move_mode** parameter always errors on service request, regardless of **ContinuousMode, RelativeMode or AbsoluteMode** setting, only the default **ContinuousMode** works.

As it works now, there is no method to do a full PAN RIGHT/LEFT or TILT UP/DOWN without executing multiple commands.  I was thinking that possibly setting the distance 1.0 would mean full span of camera,  but that is not the case.  Possibly if the distance parameter is not defined, then the camera should do full PAN/TILT direction, or another option is to map the 0.0 to 1.0 range to the actual camera min/max movement range.  Maybe this is what AbsoluteMode is to provide, but it does not work as already stated above.

Let me know if I can help any further.  I really appreciate you looking into this.  My camera has been on the bench since I got it several months ago hoping someone might get ONVIF to work.  Thanks so much!

REALogics commented:
Looking further into **move_mode** further, it is not documented correctly:

FROM: ContinuousMode, RelativeMode or AbsoluteMode setting, only the default ContinuousMode

TO: ContinuousMove, RelativeMove or AbsoluteMove setting, only the default ContinuousMove

**RelativeMove fails as follows:**
```
2020-01-02 11:07:16 DEBUG (MainThread) [custom_components.onvif.camera] Calling RelativeMove PTZ | Pan = 0 | Tilt = 0 | Zoom = 0 | Speed = 0
2020-01-02 11:07:16 DEBUG (MainThread) [custom_components.onvif.camera] Retrieved '2' profiles
2020-01-02 11:07:16 DEBUG (MainThread) [custom_components.onvif.camera] Using profile index '0'
2020-01-02 11:07:16 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.1833385872] No namespace defined for 'Ter' ('Ter:PTZNotSupported')
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 133, in handle_call_service
    connection.context(msg),
  File "/usr/src/homeassistant/homeassistant/core.py", line 1235, in async_call
    await asyncio.shield(self._execute_service(handler, service_call))
  File "/usr/src/homeassistant/homeassistant/core.py", line 1260, in _execute_service
    await handler.func(service_call)
  File "/config/custom_components/onvif/camera.py", line 128, in async_handle_ptz
    pan, tilt, zoom, distance, speed, move_mode, continuous_duration
  File "/config/custom_components/onvif/camera.py", line 416, in async_perform_ptz
    await self._ptz_service.RelativeMove(req)
  File "/usr/local/lib/python3.7/site-packages/zeep/asyncio/bindings.py", line 20, in send
    return self.process_reply(client, operation_obj, response)
  File "/usr/local/lib/python3.7/site-packages/zeep/wsdl/bindings/soap.py", line 195, in process_reply
    return self.process_error(doc, operation)
  File "/usr/local/lib/python3.7/site-packages/zeep/wsdl/bindings/soap.py", line 348, in process_error
    subcode_value_element.text, subcode_value_element.nsmap, None
  File "/usr/local/lib/python3.7/site-packages/zeep/utils.py", line 30, in as_qname
    raise XMLParseError("No namespace defined for %r (%r)" % (prefix, value))
zeep.exceptions.XMLParseError: No namespace defined for 'Ter' ('Ter:PTZNotSupported')
```
**AbsoluteMove fails as follows:**
```
2020-01-02 11:09:06 DEBUG (MainThread) [custom_components.onvif.camera] Calling AbsoluteMove PTZ | Pan = 0 | Tilt = 0 | Zoom = 0 | Speed = 0
2020-01-02 11:09:06 DEBUG (MainThread) [custom_components.onvif.camera] Retrieved '2' profiles
2020-01-02 11:09:06 DEBUG (MainThread) [custom_components.onvif.camera] Using profile index '0'
2020-01-02 11:09:06 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.1833385872] None
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 133, in handle_call_service
    connection.context(msg),
  File "/usr/src/homeassistant/homeassistant/core.py", line 1235, in async_call
    await asyncio.shield(self._execute_service(handler, service_call))
  File "/usr/src/homeassistant/homeassistant/core.py", line 1260, in _execute_service
    await handler.func(service_call)
  File "/config/custom_components/onvif/camera.py", line 128, in async_handle_ptz
    pan, tilt, zoom, distance, speed, move_mode, continuous_duration
  File "/config/custom_components/onvif/camera.py", line 426, in async_perform_ptz
    await self._ptz_service.AbsoluteMove(req)
  File "/usr/local/lib/python3.7/site-packages/zeep/asyncio/bindings.py", line 13, in send
    options["address"], envelope, http_headers
  File "/usr/local/lib/python3.7/site-packages/zeep/asyncio/transport.py", line 107, in post_xml
    response = await self.post(address, message, headers)
  File "/usr/local/lib/python3.7/site-packages/zeep/asyncio/transport.py", line 95, in post
    proxy=self.proxy,
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client.py", line 504, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 847, in start
    message, payload = await self._protocol.read()  # type: ignore  # noqa
  File "/usr/local/lib/python3.7/site-packages/aiohttp/streams.py", line 591, in read
    await self._waiter
aiohttp.client_exceptions.ServerDisconnectedError: None
```

djpremier commented on homeassistant/components/onvif/camera.py:
```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
-                await self._ptz_service.ContinuousMove(req)
+                req = self._ptz_service.create_type(move_mode)
+                req.ProfileToken = await self.async_obtain_profile_token()
+                if move_mode == "ContinuousMove":
```
I recommend using the constant you defined instead of the raw string.

djpremier commented on homeassistant/components/onvif/camera.py:
```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
+                    req = self._ptz_service.create_type("Stop")
+                    req.ProfileToken = await self.async_obtain_profile_token()
+                    await self._ptz_service.Stop({"ProfileToken": req.ProfileToken})
+                elif move_mode == "RelativeMove":
```
I recommend using the constant you defined instead of the raw string.

olijouve replied:
You're absolutely right, thanks.
It's fixed

djpremier commented on homeassistant/components/onvif/camera.py:
```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
+                        "Zoom": {"x": speed_val},
+                    }
+                    await self._ptz_service.RelativeMove(req)
+                elif move_mode == "AbsoluteMove":
```
I recommend using the constant you defined instead of the raw string.

REALogics replied:
Thanks @djpremier

What I was pointing out in my previous post is that the description/documentation additions for **move_mode** in camera/services.yaml does not match **move_mode** in onvif/camera.py code.  EG: services.yaml "AbsoluteMode" verses camera.py "AbsolueMove"

Even so, setting move_mode to AbsoluteMove or RelativeMove **does not work** for my FOSCAM ONVIF camera.  The resulting logs were included in my previous post.

REALogics commented:
Logging would be more meaningful if %d were changed to %4.2f for camera.py line 385:
```
            _LOGGER.debug(
                "Calling %s PTZ | Pan = %4.2f | Tilt = %4.2f | Zoom = %4.2f | Speed = %4.2f",
                move_mode,
                pan_val,
                tilt_val,
                zoom_val,
                speed_val,
            )
```

olijouve committed:
Update services.yaml

fix wrong wording for move_mode

olijouve commented:
@REALogics, thanks for your feedback. I've fixed bad documentation on move_mode.
About non functionnal features you've experimented and reported i think it is because this PR is about ONVIF PTZ wsdl ver10 support and your camera is ONVIF 2.2 so PTZ wsdl ver20.

Originally I was only looking to be able to move poor chinese Goke firmware based cameras.

I tried to find an ONVIF PTZ simulator/server to see how i could test and implement pztz/ver20/wsdl calls but i was in vain...
I absolutely not need it as my wife forbid me to add more cams in house :/ but i've just ordered a (ouch expensive !) Foscam R4 cam that is supposed to be ONVIF 2.2 with zoom support :p 

I'll do what i can within this month to see how i can add support for it as my first intent is only for dev purpose then to send it back to Am.zon(not final if camera is really good and w/ successfull support)

Also i think i'll switch to a custom_component integration for the developpement stage, it seems more suitable.

olijouve committed:
Update camera.py

Using defined constants instead of  raw strings in conditions

olijouve committed:
Update camera.py

Replace integer to floating point in logger debug PTZ values

olijouve commented:
> Logging would be more meaningful if %d were changed to %4.2f for camera.py line 385:

It's done

REALogics commented:
Awesome @olijouve.  Thanks for working on this!  I really do appreciate it.

olijouve commented:
> All of my camera functions work using the "[ONVIF Device Manager](https://sourceforge.net/projects/onvifdm/)" program

@REALogics, i received the foscam R4 zoom x8 2 days ago but even after update it to last firmware i have still a poor support on ODM compare to what you said w/ your camera.
For me RelativeMove doesn't work at all, there's no support for it. Zoom in Continuous/Absolute doesn't work neither.

Could you tell me what firmware version you've got on your camera, and could you share me some screenshots of your whole profiles (media+ ptz)(some sub nodes may need to be uncolapsed).

I think Foscam left something wrong/incomplete in default ptz profile that do not allow R4 to operates zoom(default pan, tilt, zoom speed all set to 0, and default zoom spaces are unreconnized !). I also have good looking ptz sub node profile but it is unusable...

The more you can give me on it, the best as i intend to write to foscam support on this.

Now the interessing part about dev, so far using Zeep module in a standalone python script here is what i've run successfully over that foscam R4:
- ContinuousMove (no Zoom and Timeout support)
- AbsoluteMove (no Zoom)
- GotoHomePosition ( SetHomePosition doesn't work as home position cannot be overwrited)
- GetPresets
- GotoPreset
- SetPreset (awesome for those last 3)

REALogics commented:
Following are versions given when logging in IP directly to camera FI9928P:
```
System Firmware Version: 1.11.1.11
Application Firmware Version: 2.74.1.77
Plug-in Version 5.1.0.13
```
There are a lot of setting with drop down menus, but I do not see anything named "profile".  I've attached the camera manual here.  Possibly you can take a look and tell the settings you would like to know.

[User Manual for FI9928P V1.0.6_English.pdf](https://github.com/home-assistant/home-assistant/files/4028371/User.Manual.for.FI9928P.V1.0.6_English.pdf)

olijouve commented:
Thanks @REALogics, I am on same firmware version and the user manual si the same for my cam expect for first pages showing visuals of the camera.

For profile i was meant for it in Onvif Device Manager(ODM) because i am surprise you said all your camera functions are working in ODM but not the case for my camera which has the same firmware. Can you confirm you are able to operate Zoom in ODM (in Continuous/Absolute mode ? )

See attachement for ODM profile screenshot: 
<img width="1294" alt="Capture d’écran 2020-01-07 à 02 53 11" src="https://user-images.githubusercontent.com/17448560/71862465-41ac7580-30fa-11ea-891a-916e30240877.png">

Would be nice if you could send me that profile for your camera.

I've stared to implement preset and homing in an onvif custom_custom component, full pan/tilt too.

jkoenig72 commented:
Hi,

nice work @olijouve and the others!

I just copied this over the original file within the HA install location. Works. Only thing: it does a complete zoom in and out, would be better it does for 1 or 2 sec and then stops and wait for the next zoom in or out.

I had the same issue when I was trying to script this directly in

https://github.com/FalkTannhaeuser/python-onvif-zeep

Same problem there. By using the continuous mode, it does not stop even when triggering

ptz.Stop({'ProfileToken': request.ProfileToken}) or using your timeout:

I never get that working. Now you can say maybe this cam does not implement is correctly? Maybe but it works when using Zoneminder's onvif netcat "Auto Stop Timeout" setting in the Zoneminder UI... wondering what netcat does differently than python-onvif...

Any idea what I can try to get ContinuousMode working with not completly ZOOM in / out?

Thanks!

/Joerg

REALogics commented:
@olijouve

Yes I was able to zoom in/out with ODM.  I have not figured out how to use ODM to access camera remotely over internet, so I can't capture the info you requested until I return home this evening EST.  I will update then.

REALogics commented:
@olijouve camera profile attached:

![ODM1](https://user-images.githubusercontent.com/7430727/71939381-6fd49880-3180-11ea-8dfb-f455045c668a.PNG)
![ODM2](https://user-images.githubusercontent.com/7430727/71939418-8da1fd80-3180-11ea-9a43-d93a16bf7e42.PNG)

REALogics commented:
Using ODM Absolute/Relative/Continuous moves at bottom right of ODM program window, I am not able to execute ZOOM for any of the three move types.

![Relative](https://user-images.githubusercontent.com/7430727/71940252-a3fd8880-3183-11ea-891d-2c7b8b6db03f.PNG)
![Absolute](https://user-images.githubusercontent.com/7430727/71940253-a3fd8880-3183-11ea-9b6c-47a256502385.PNG)
![Continuous](https://user-images.githubusercontent.com/7430727/71940254-a3fd8880-3183-11ea-816e-9c7737a18b0d.PNG)

olijouve commented:
@REALogics
> **Sorry, I was wrong with ZOOM. It DOES NOT work with ODM**

Arf, same R4 firmware, same unsupported operations. But I'm perplex as i can't operate RelativeMove on mine, would you mind to recheck that too ?

olijouve commented:
> Same problem there. By using the continuous mode, it does not stop even when triggering
> 
> ptz.Stop({'ProfileToken': request.ProfileToken}) or using your timeout:
> 
> I never get that working. Now you can say maybe this cam does not implement is correctly? Maybe but it works when using Zoneminder's onvif netcat "Auto Stop Timeout" setting in the Zoneminder UI... wondering what netcat does differently than python-onvif...
> 
> Any idea what I can try to get ContinuousMode working with not completly ZOOM in / out?

@jkoenig72, yes there is a way.
Stop must be called with pan/tilt or zoom parameter to indicate what should be stopped. See [onvif specs](https://www.onvif.org/specs/srv/ptz/ONVIF-PTZ-Service-Spec-v221.pdf) for more details.
However this should have better support in the new onvif custom_component i'm rewriting.
Stay tuned.

olijouve commented:
@REALogics 
> Assuming FOSCAM does not offer a fix for this, it would be nice if HA could access presets names stored in the camera configured by the user. I am able to set pan/tilt presets and then execute them. That way you could set preset positions like a front door, mailbox, garage etc... presets.
> 

As previously said i'm able to do it and since last night it's implemented in my new onvif2 custom component  :smiley: 
Unpublished yet, should be able to have a release within few days. Stay tuned

REALogics commented:
@olijouve RE:ODM

I must apologize, being a newbie at this...  over the past months I have been confused between using ODM verses Camera Service via http.  In addition, I really understood how ODM settings worked.  If I open ODM and click on Setting->Relative, the ODM screen changes from double >> arrows to single > arrow for up down left right movement.  When I click on any of the arrows, I get the same response:

`Server returned an invalid SOAP fault. Please see InnerException for more details.`

When I press on the -+ zoom buttons, nothing happens and no error message.  

The screen shot supplied in earlier post for RelativeMove presets, the **Zoom** and **Zoom Speed** sliders are not there, only a value of 0 is give for each... so you can't set them.  There is also no slider for ContinuousMove zoom velocity parameter, only a value of 0 is showing.

It seems that in creating a RelativeMove preset, it makes no difference what values you set, since activating the Relative preset always takes you to the same position.  Not sure whats happening here.

Speed and velocity parameter settings don't seem to work at all for any move.

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -105,7 +124,9 @@
+            )
 
     hass.services.async_register(
         DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
```
Please register the custom service under the onvif domain instead of the camera domain. Custom services must be registered under the integration domain.

olijouve replied:
To adress your comment, do i need to remove this import:
`from homeassistant.components.camera.const import DOMAIN`
and replace it after imports by:
`DOMAIN="onvif"`

Is it correct ?

olijouve replied:
I've tested it that way and it did the trick. 
Just ignore my last comment if your ok with that way to do.

MartinHjelmare commented on homeassistant/components/camera/services.yaml:
```diff
@@ -69,17 +69,33 @@ record:
       example: 4
 
 onvif_ptz:
```
Please move this description to a new services.yaml file under the onvif integration package.

olijouve replied:
it is what i had made in the first place but it had no effect, but maybe related to your previous comment about wrong domain usage

olijouve replied:
Yes it was domain related, many thanks pointed it out to me !

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
+                    }
+
+                    await self._ptz_service.ContinuousMove(req)
+                    sleep(continuous_duration)
```
Sync sleep isn't allowed in the event loop, inside a coroutine. Please await `asyncio.sleep` instead.

olijouve replied:
Thanks !

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -95,6 +110,10 @@
         pan = service.data.get(ATTR_PAN, None)
         tilt = service.data.get(ATTR_TILT, None)
         zoom = service.data.get(ATTR_ZOOM, None)
+        distance = service.data.get(ATTR_DISTANCE, None)
```
Please use `dict[key]` for required schema keys and keys with default schema values.

olijouve replied:
I don't figure out what that mean yet, will try to check against other components

MartinHjelmare replied:
Like this:
```suggestion
        distance = service.data[ATTR_DISTANCE]
```

olijouve replied:
Thank you

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
             _LOGGER.error("Couldn't setup camera '%s'. Error: %s", self._name, err)
 
-    def setup_ptz(self):
+    async def setup_ptz(self):
```
Why change here to a coroutine?

olijouve replied:
Indeed, did that too quickly after reading a post in issue #27744 saying it solved an issue.
I will remove it if unecessary.

Thank you for all your comments, actually i'm rewriting a new onvif2 component which will have more complete support on other PTZ operations, i will address your change request in it.
I was thinking first to release it as a custom_component before to make a new PR when it will be widely tested and mature. This way I would not consume your time day after day on code review.
If you, Team guys, also think it would be a better way, i might better close that PR then.

MartinHjelmare replied:
Further changes to this integration are welcome but not in this PR. Please make a new separate PR for further changes.

REALogics replied:
@olijouve 

> actually i'm rewriting a new onvif2 component which will have more complete support on other PTZ operations, i will address your change request in it.

One change which might make onvif2 version more flexible would be to define PTZ parameters as floats, rather than the hard coded UP/DOWN/LEFT/RIGHT/IN/OUT.  And possibly second to retrieve parameter min/max ranges via camera ONVIF request.

> release it as a custom_component 

I am not a HA coder (at least not yet :-).  And I have been copying your GIT code to custom_component to test, all along.  I don't know yet how to do it otherwise.  So this method works fine for me.

olijouve replied:
> Why change here to a coroutine?

@MartinHjelmare almost all others methods of the OnvifCamera class are async coroutines they should not be ?

MartinHjelmare replied:
We only want to define coroutines when we need to await something inside the function. There's an overhead time wise to await a coroutine versus calling a normal function. So we try to be conservative.

REALogics replied:
@MartinHjelmare 

I am a C & C++ embedded developer, but not on HA or Python developer.

Just curious... it seems a local function is synchronous and blocking... a coroutine is asynchronous and non-blocking?  When a function requires command & response to a remote device like a camera, wouldn't that be valid for coroutine (non-blocking) function.  Or is the integration (camera/onvif) already executing in a coroutine?

MartinHjelmare replied:
It depends on the device library if it supports asyncio or not and what api it exposes.

olijouve committed:
Update services.yaml

olijouve committed:
Update services.yaml

olijouve committed:
Update camera.py

REALogics commented:
@olijouve 

Was wondering if you had tried connecting to FOSCAM regarding ONVIF zoom issues?

olijouve commented:
> @olijouve
> 
> Was wondering if you had tried connecting to FOSCAM regarding ONVIF zoom issues?

@REALogics, I did, after 3 emails exchanged to says it was the only missing onvif function here are their last words(I wish other foscam devices would write their support too):

Hi olivier Jouve,

Thanks for contacting Foscam support!
Sorry, our camera's Onvif protocol does not support the PTZ control.
Thanks.



Waiting for your kindly reply!

Best Regards,

Zion Zhang
Customer Service Representative
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Your satisfaction is our goal. Any feedback will be highly valued.
If you have any suggestion to our service or product, please e-mail feedback@foscam.com.

REALogics commented:
@olijouve 

So much for 

> Your satisfaction is our goal.

I have also followed up with email to FOSCAM to get an answer... at least to gig them on their documentation that states their PTZ camera is ONVIF compatible... yet documentation has  no disclaimers on zoom, or provides an ONVIF compatibility matrix.

olijouve commented:
> @olijouve
> 
> So much for
> 
> > Your satisfaction is our goal.
> 
> I have also followed up with email to FOSCAM to get an answer... at least to gig them on their documentation that states their PTZ camera is ONVIF compatible... yet documentation has no disclaimers on zoom, or provides an ONVIF compatibility matrix.

@REALogics, I didn't reply after that as I was able get zoom to work in a different way.
Quick foscam UI javascript reading and deduction lead me to this:

curl "http://<HOST_IP>:88/cgi-bin/CGIProxy.fcgi?usr=<USERNAME>&pwd=<PASSWORD>&cmd=zoomIn"

curl "http://<HOST_IP>:88/cgi-bin/CGIProxy.fcgi?usr=<USERNAME>&pwd=<PASSWORD>&cmd=zoomOut"

curl "http://<HOST_IP>:88/cgi-bin/CGIProxy.fcgi?usr=foscam&pwd=pwd4iotcam&cmd=zoomStop"

You can give it a try it's working nice, but this will be related to Foscam integration, not ONVIF.

olijouve committed:
Update camera.py

REALogics commented:
@olijouve 

I could not find CGI included in the specs of either the Foscam R4 or FI9928P camera, but its worth a try.  I guess you are keeping your camera... now :-)

I have tested your most recent camera and onvif component changes for @MartinHjelmare, and PT works as before.

I appreciate all you have done pursuing PTZ on FOSCAM.  Thanks.

REALogics commented:
@olijouve 

You lost the changes on services.yaml...

**NO**: ContinuousMode, RelativeMode or AbsoluteMode | ContinuousMode

**YES**: ContinuousMove, RelativeMove or AbsoluteMove | ContinuousMove

olijouve commented:
> @olijouve
> 
> You lost the changes on services.yaml...
> 
> **NO**: ContinuousMode, RelativeMode or AbsoluteMode | ContinuousMode
> 
> **YES**: ContinuousMove, RelativeMove or AbsoluteMove | ContinuousMove

I don't see it:
https://github.com/home-assistant/home-assistant/pull/30152/files#diff-91da43527b1dbfd50ec898e66e128d75

olijouve commented:
>  Seems energy should be put into the "foscam" component instead of "onvif" component.

As everything else already exist, there is only 3 new fonctions to add in the pypi libpyfoscam project(https://pypi.org/project/libpyfoscam/).
I've just added them, we will still need minor changes in HA foscam integration. I'm waiting for libfoscam to merge my PR to move on HA Foscam part.

And indeed i might keep the camera now ;)

REALogics commented:
@olijouve

Regarding services.py, I copied from your GIT components folder both camera and onvif to my  custom_components folder, then restarted RPi and tested.  The text shows up in Developer Tools->Services->Call Service example.  Is this a proper method for testing changes?

olijouve commented:
@REALogics there are both services.yaml in this PR and camera/services.yaml was reverted to its original state, please apply that change too. No need rpi reboot, simple HA restart is enough

REALogics commented:
@olijouve 

I deleted camera and onvif from my custom_components directory, pasted your dev branch onvif into my custom_components and tested.  It all looks and plays as I understand it should.

olijouve commented:
@REALogics I still need more time to work on it and to write documentation before i can make a new PR but as the early adopter you are you should try this, it also supports Presets and advanced move mode that permit to set position with "vector" values : 

https://github.com/olijouve/onvif2

You will need new config params:
`camera:
  - platform: onvif
    host: 192.168.1.16
    port: 888
    name: foscam
    username: "USERNAME"
    password: "PASSWORD"
    profile: 1 # lowest compression lowest lag
    continuous_timeout_compliance: False 
    rtsp_transport: "RTSP"
`
Also for Foscam i succeed to get a low ~2s lag between a ptz cmd and the visible motion.
Less compression give less lag, here is my video config in Foscam web ui.

<img width="730" alt="Capture d’écran 2020-01-14 à 03 08 03" src="https://user-images.githubusercontent.com/17448560/72566308-dc077880-38b3-11ea-9cfb-1cbc2075a79c.png">

Note : Foscam only permit 6 presets and removing preset is not supported in their onvif implementation.

REALogics commented:
@olijouve 

Awesome!  Probably not today... but I will give it a try and let you know.

MartinHjelmare commented:
I don't believe all my comments have been resolved. I don't changes related to them.

olijouve commented:
> I don't believe all my comments have been resolved. I don't changes related to them.
Sorry,thought the change request were only attached to services.yaml comment...

So indeed i didn't, but I've adressed it all in a more advanced version available bellow(+ @balloob comment found on an other PR concerning undesirable all cameras selection if no entity_id is specified) : 
https://github.com/olijouve/onvif2

I should close that PR and open a new one, i'll do it when i'll be able to find time to update related documentation.

MartinHjelmare commented:
My comments in this PR need to be addressed before we can merge this PR.

olijouve committed:
use dict[key] for required schema keys and keys with default schema values

olijouve committed:
remove async for setup_ptz method

olijouve committed:
lint error

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -48,20 +48,30 @@
+CONTINUOUS_MOVE = "ContinuousMove"
+RELATIVE_MOVE = "RelativeMove"
+ABSOLUTE_MOVE = "AbsoluteMove"
 PTZ_NONE = "NONE"
```
Is the value `"NONE"` used by the device library? Otherwise we can remove it and use `dict.get` to get the config item. Then the default value will be `None`.

olijouve committed:
remove unecessary PTZ_NONE = "NONE"

changed request by @MartinHjelmare

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -82,22 +80,22 @@
-        vol.Optional(ATTR_PAN, default=PTZ_NONE): vol.In([DIR_LEFT, DIR_RIGHT, PTZ_NONE]),
-        vol.Optional(ATTR_TILT, default=PTZ_NONE): vol.In([DIR_UP, DIR_DOWN, PTZ_NONE]),
-        vol.Optional(ATTR_ZOOM, default=PTZ_NONE): vol.In([ZOOM_OUT, ZOOM_IN, PTZ_NONE]),
+        vol.Optional(ATTR_PAN, default=None): vol.In([DIR_LEFT, DIR_RIGHT, None]),
```
Don't set a default at all. And remove `None` from the list of accepted values.

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -92,9 +106,13 @@
-        pan = service.data.get(ATTR_PAN, None)
-        tilt = service.data.get(ATTR_TILT, None)
-        zoom = service.data.get(ATTR_ZOOM, None)
+        pan = service.data[ATTR_PAN]
```
```py
pan = service.data.get(ATTR_PAN)
```

MartinHjelmare replied:
```suggestion
        pan = service.data.get(ATTR_PAN)
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -307,37 +356,73 @@ def __init__(self, hass, config):
-                }
-            }
+            pan_val = (
+                distance if pan == DIR_RIGHT else -distance if pan == DIR_LEFT else 0
```
Instead of this long if block we can make a map between the pan direction and the factor to multiply the distance.

```py
PAN_FACTOR = {
    DIR_RIGHT: 1,
    DIR_LEFT: - 1,
}

...

pan_val = distance * PAN_FACTOR.get(pan, 0)
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -79,9 +87,15 @@
+        vol.Optional(ATTR_TILT, default=None): vol.In([DIR_UP, DIR_DOWN, None]),
+        vol.Optional(ATTR_ZOOM, default=None): vol.In([ZOOM_OUT, ZOOM_IN, None]),
+        ATTR_MOVE_MODE: vol.In(
+            [CONTINUOUS_MOVE, RELATIVE_MOVE, ABSOLUTE_MOVE, None]
```
Please remove `None` here too.

MartinHjelmare reviewed (changes requested):
Please update tilt and zoom too as described for pan below.

olijouve committed:
addressing @ MartinHjelmare comments

- Remove None in defaluts and dicts
- Replace long if blocks

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -371,15 +371,21 @@ def setup_ptz(self):
-            zoom_val = (
-                distance if zoom == ZOOM_IN else -distance if zoom == ZOOM_OUT else 0
-            )
+            PAN_FACTOR = {
```
We can move these constants out to the module level.

MartinHjelmare commented on homeassistant/components/onvif/services.yaml:
```diff
@@ -2,15 +2,30 @@ onvif_ptz:
     tilt:
-      description: 'Tilt direction. Allowed values: UP, DOWN, NONE'
-      example: 'UP'
+      description: "Tilt direction. Allowed values: UP, DOWN, NONE"
```
Remove NONE.

olijouve committed:
remove NONE

olijouve committed:
lint issue

olijouve committed:
Update camera.py

olijouve commented:
Hello @MartinHjelmare, i'm stucked with a lint issue:
`homeassistant/components/onvif/services.yaml:25: stoping  ==> stopping`

I absolutely do not see what's wrong in line 25 of that services.yaml file that cause that codespell error.

MartinHjelmare commented on homeassistant/components/onvif/services.yaml:
```diff
@@ -2,15 +2,30 @@ onvif_ptz:
+      default: 0.5
+      example: 0.5
+    continuous_duration:
+      description: "Set ContinuousMove delay in seconds before stoping the move"
```
It's just a typo. Change to "stopping".

olijouve replied:
Thanks !

olijouve committed:
Fix lint error - typo

olijouve commented:
@MartinHjelmare, i think i've adressed every changes you asked for.

MartinHjelmare commented:
Please also add a breaking change paragraph about moving the services from the camera domain to the onvif domain. 

Briefly describe what has changed and what the user needs to do to cope with the breaking change.

olijouve commented:
@MartinHjelmare, i've been very busy thoses last days. 
Does my description update is clear/complete enough ?

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -48,20 +47,31 @@
+RELATIVE_MOVE = "RelativeMove"
+ABSOLUTE_MOVE = "AbsoluteMove"
 
 SERVICE_PTZ = "onvif_ptz"
```
Please rename the service to just `ptz`.

MartinHjelmare reviewed (commented):
Please rename the service from `onvif_ptz` to `ptz`. Make sure to update the services.yaml file too.

olijouve committed:
rename onvif_ptz service to just ptz

olijouve committed:
rename onvif_ptz service to just ptz

olijouve commented:
home-assistant.io onvif documentation was also been updated accordingly

MartinHjelmare commented:
Good! Then it's just the comments on dictionary access above, remaining.

olijouve commented:
> Good! Then it's just the comments on dictionary access above, remaining.

Done

MartinHjelmare commented:
The comments need to be addressed by pushing new commits fixing the dictionary access.

olijouve committed:
use dict[key] when default values are set

use service.data[key] instead of service.data.get[key] when default value is set in service schema

olijouve commented:
> The comments need to be addressed by pushing new commits fixing the dictionary access.

Sorry, i thought i already did it but not (i only did it in my newer version of this component)

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -92,9 +106,13 @@
-        pan = service.data.get(ATTR_PAN, None)
-        tilt = service.data.get(ATTR_TILT, None)
-        zoom = service.data.get(ATTR_ZOOM, None)
+        pan = service.data.get[ATTR_PAN]
```
Please fix this too.

olijouve committed:
adresse comment: use dict[key] for pan tilt zoom

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -79,9 +89,13 @@
-        ATTR_PAN: vol.In([DIR_LEFT, DIR_RIGHT, PTZ_NONE]),
-        ATTR_TILT: vol.In([DIR_UP, DIR_DOWN, PTZ_NONE]),
-        ATTR_ZOOM: vol.In([ZOOM_OUT, ZOOM_IN, PTZ_NONE]),
+        vol.Optional(ATTR_PAN): vol.In([DIR_LEFT, DIR_RIGHT, None]),
```
We don't want to accept `None` as a valid value.
```suggestion
        vol.Optional(ATTR_PAN): vol.In([DIR_LEFT, DIR_RIGHT]),
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -92,9 +106,13 @@
-        tilt = service.data.get(ATTR_TILT, None)
-        zoom = service.data.get(ATTR_ZOOM, None)
+        pan = service.data[ATTR_PAN]
+        tilt = service.data[ATTR_TILT]
```
```suggestion
        tilt = service.data.get(ATTR_TILT)
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -92,9 +106,13 @@
-        zoom = service.data.get(ATTR_ZOOM, None)
+        pan = service.data[ATTR_PAN]
+        tilt = service.data[ATTR_TILT]
+        zoom = service.data[ATTR_ZOOM]
```
```suggestion
        zoom = service.data.get(ATTR_ZOOM)
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -79,9 +89,13 @@
-        ATTR_ZOOM: vol.In([ZOOM_OUT, ZOOM_IN, PTZ_NONE]),
+        vol.Optional(ATTR_PAN): vol.In([DIR_LEFT, DIR_RIGHT, None]),
+        vol.Optional(ATTR_TILT): vol.In([DIR_UP, DIR_DOWN, None]),
+        vol.Optional(ATTR_ZOOM): vol.In([ZOOM_OUT, ZOOM_IN, None]),
```
```suggestion
        vol.Optional(ATTR_TILT): vol.In([DIR_UP, DIR_DOWN]),
        vol.Optional(ATTR_ZOOM): vol.In([ZOOM_OUT, ZOOM_IN]),
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -92,9 +106,13 @@
+        zoom = service.data[ATTR_ZOOM]
+        distance = service.data[ATTR_DISTANCE]
+        speed = service.data[ATTR_SPEED]
+        move_mode = service.data[ATTR_MOVE_MODE]
```
```suggestion
        move_mode = service.data.get(ATTR_MOVE_MODE)
```

MartinHjelmare committed:
Apply suggestions from code review

MartinHjelmare reviewed (approved):
Please confirm that the latest commit works. Then we can merge.

olijouve commented:
> Please confirm that the latest commit works. Then we can merge.
Yes i confirm it works for me.
